### PR TITLE
Call dlclose on libgdiplusHandle

### PR DESCRIPTION
**Pull request by kasper3**
Related discussion at https://github.com/dotnet/corefx/issues/29426#issuecomment-385717046.

### CODE REVIEW

File: src/System.Drawing.Common/tests/Helpers.cs [hunk 1]
```diff
@@ -27,21 +27,24 @@ public static bool GetGdiplusIsAvailable()
...
+                    libgdiplusHandle = dlopen("libgdiplus.so", RTLD_NOW);
+                    if (libgdiplusHandle == IntPtr.Zero)
                     {
-                        nativeLib = dlopen("libgdiplus.so.0", RTLD_NOW);
+                        libgdiplusHandle = dlopen("libgdiplus.so.0", RTLD_NOW);
                     }
                 }
 
-                return nativeLib != IntPtr.Zero;
+                int dlcloseResult = dlclose(libgdiplusHandle);
```
Comment (hughbe) [inline src/System.Drawing.Common/tests/Helpers.cs]: Not necessarily correct: please change this to

```
if(libgdiplusHandle != IntPtr.Zero)
{
  dlclose(libgdiplusHandle);
}
```

Comment (kasper3) [reply src/System.Drawing.Common/tests/Helpers.cs]: Makes sense, changed to close conditionally.

File: src/System.Drawing.Common/tests/Helpers.cs [hunk 2]
```diff
@@ -27,21 +27,26 @@ public static bool GetGdiplusIsAvailable()
...
+                    libgdiplusHandle = dlopen("libgdiplus.so", RTLD_NOW);
+                    if (libgdiplusHandle == IntPtr.Zero)
                     {
-                        nativeLib = dlopen("libgdiplus.so.0", RTLD_NOW);
+                        libgdiplusHandle = dlopen("libgdiplus.so.0", RTLD_NOW);
                     }
                 }
 
-                return nativeLib != IntPtr.Zero;
+                if(libgdiplusHandle != IntPtr.Zero)
```
Comment (danmosemsft) [inline src/System.Drawing.Common/tests/Helpers.cs]: Nit, space before open paren please

**Comment by kasper3** [comment]
In Fedora (only), it seems like `dlopen(libgdiplus)` seems to be lazy initializing some glib resources and aborting on `dlclose()`. In my little investigation, it seems like error is coming from https://github.com/GNOME/glib/blob/79d9ea2598c1385104b4c88843d5d92b0203c498/glib/gthread-posix.c#L1012. libgdiplus depends on glib.

https://mc.dot.net/#/user/kasper3/pr~2Fjenkins~2Fdotnet~2Fcorefx~2Fmaster~2F/test~2Ffunctional~2Fcli~2F/98e47bce973c2f1ef1f3534124c0d0e72bf1888f/workItem/System.Drawing.Common.Tests/wilogs

> 2018-05-02 07:37:55,013: INFO: proc(54): run_and_log_output: Output: GLib (gthread-posix.c): Unexpected error from C library during 'pthread_key_create': Resource temporarily unavailable.  Aborting.

**Comment by qmfrederik** [comment]
@kasper3 What problem is this trying to fix, though? System.Drawing.Common will load libgdiplus and never unload it (when found) so the tests aren't leaking any resource (except for incrementing a counter).

I'm not sure libgdiplus and its dependencies were designed/tested to be loaded & unloaded.

`GetGdiplusIsAvailable` could cache the result, though, and that would avoid frequently calling `dlopen`.

**Comment by kasper3** [comment]
@qmfrederik, general principle is: if you allocate it (memory), you deallocate it. If you open it (file), you close it.

**Comment by qmfrederik** [comment]
Yes, I understand that.

However, in this case System.Drawing.Common will load libgdiplus and keep it alive. Although the unit tests are calling `dlopen` on libgdiplus, only one copy of libgdiplus will be loaded, and that copy will be always be there because of System.Drawing.Common.

What this change will do on a system where libgdiplus is installed (based on my understanding):
1. The first call to `GetGdiplusIsAvailable` will load and unload libgdiplus
2. The unit test for which `GetGdiplusIsAvailable` was called will call into System.Drawing.Common and libgdiplus will be loaded again. System.Drawing.Common never calls `dlclose`.
3. Subsequent calls to `GetGdiPlusAvailable` will call `dlopen` and `dlclose`; this will increment/decrement the reference count to libgdiplus but it will never unload libgdiplus.

Step 1 introduces a load/unload cycle of libgdiplus, something which I doubt libgdiplus has been designed for and something which will not normally happen for code using System.Drawing.Common.

It is step 1 which is causing the CI to fail on Fedora.

But I do think that frequent calls to `dlopen` are overkill, so perhaps having `IsLibGdiPlusAvailable` cache the result is a better option?

**Comment by kasper3** [comment]
> Subsequent calls to GetGdiPlusAvailable will call dlopen and dlclose; this will increment/decrement the reference count to libgdiplus but it will never unload libgdiplus.

Yes, except if reference count approaches 0 by dlopen, then lib gets unloaded from memory.
We can cache the result, since we only care about availability. I have noticed that xunit initializes classes multiple times, so for caching, should we just change:

```diff
- public static bool CanTestSetHdevmode_IntPtr_Success =>
+  public static bool CanTestSetHdevmode_IntPtr_Success =
```

at

https://github.com/dotnet/corefx/blob/2d7b6dc/src/System.Drawing.Common/tests/Printing/PrinterSettingsTests.cs#L596 ?

**Comment by stephentoub** [comment]
@safern, should this PR be merged? Closed? Updated based on additional feedback?

**Comment by jkotas** [comment]
> In Fedora (only), dlopen(libgdiplus) seems to be lazy initializing some glib resources and aborting on dlclose(). 

`libgdiplus` is clearly not expected to be unloaded. This PR should be closed. 

`dlopen` / `dlclose` can be only used for libraries that are designed to be unloadedable. It is fraction of the libraries out there. For example, libcoreclr is not designed to be unloadable either.

**Comment by safern** [comment]
I agree.

**Pull request closed without merging**